crates/bevy_ecs/src/query/state.rs

```diff
@@ -21,6 +21,17 @@ use super::{
     QuerySingleError, ROQueryItem,
 };
 
+/// An ID for either a table or an archetype. Used for Query iteration.
```

Comment:
I don't know if it would be useful to specify that this is used for optimizing query iteration; in the case where all components are in tables, we can iterate through table_ids directly instead of archetypes

Reply:
I would also maybe add a comment on why this being a `union` is ok here, as unions are rarer than enums. It's because we know which variant to use depending on `QueryState::IS_DENSE` so we can skip storing the variant type?